File: google/backtracking/partition-to-k-equal-sum-subsets.js
/* 
Time: O(k*2^N)
k: subset count
Space: O(N)
Recursive stack : O(N) and visit: O(N)
*/
var canPartitionKSubsets = function (nums, k) {
    const totalSum = nums.reduce((sum, val) => sum + val, 0);
    if (totalSum % k) return false;
    const target = totalSum / k;
    const visit = new Array(nums.length).fill(false);
    nums.sort((a, b) => b - a);

    function backtrack(idx, k, sum) {
        if (k === 0) return true;
        if (sum === target) return backtrack(0, k - 1, 0);

        for (let i = idx; i < nums.length; i++) {
            if (visit[i] || (sum + nums[i]) > target) continue;
            visit[i] = true
            if (backtrack(i + 1, k, sum + nums[i])) return true;
            visit[i] = false;
        }
        return false;
    }
    return backtrack(0, k, 0);
};

File: masteringDsAlgo/src/dsAlgo/backtracking/PartitionToKEqualSumSubsets.java
package dsAlgo.backtracking;

import java.util.Arrays;

public class PartitionToKEqualSumSubsets {
    private boolean backtrack(int idx, int[] arr, int k, int target, int sum, boolean[] used) {
        if (k == 0) return true;
        if (target == sum) return backtrack(0, arr, k - 1, target, 0, used);

        for (int i = idx; i < arr.length; i++) {
            if (used[i] || sum + arr[i] > target) continue;
            used[i] = true;
            if (backtrack(i + 1, arr, k, target, sum + arr[i], used)) return true;
            used[i] = false;
        }

        return false;
    }
    void reverse(int[] arr) {
        for (int i = 0, j = arr.length - 1; i < j; i++, j--) {
            int tmp = arr[i];
            arr[i] = arr[j];
            arr[j] = tmp;
        }
    }
    public boolean canPartitionKSubsets(int[] arr, int k) {
        int totalSum = Arrays.stream(arr).sum();
        if (totalSum % k != 0) return false;
        int target = totalSum / k;
        boolean[] used = new boolean[arr.length];
        Arrays.fill(used, false);
        Arrays.sort(arr);
        reverse(arr);
        return backtrack(0, arr, k, target, 0, used);
    }
}
